beater/java_attacher/java_attacher.go

```diff
@@ -22,61 +22,389 @@ import (
-			cfg.JavaBin = filepath.Join(jh, "/bin/java")
-		} else {
-			bin, err := exec.LookPath("java")
+func (j *JavaAttacher) addDiscoveryRule(rule DiscoveryRule) {
+	j.discoveryRules = append(j.discoveryRules, rule)
+	j.logger.Debugf("added discovery rule: %v", rule)
+}
+
+func New(cfg config.JavaAttacherConfig) JavaAttacher {
+	logger := logp.NewLogger("java-attacher")
+	attacher := JavaAttacher{
+		logger:               logger,
+		enabled:              cfg.Enabled,
+		agentConfigs:         cfg.Config,
+		downloadAgentVersion: cfg.DownloadAgentVersion,
+		rawDiscoveryRules:    cfg.DiscoveryRules,
+	}
+	for _, flag := range cfg.DiscoveryRules {
+		for name, value := range flag {
+			switch name {
+			case "include-all":
+				attacher.addDiscoveryRule(IncludeAllRule{})
+			case "include-user":
+				attacher.addDiscoveryRule(UserDiscoveryRule{user: value, isIncludeRule: true})
+			case "exclude-user":
+				attacher.addDiscoveryRule(UserDiscoveryRule{user: value, isIncludeRule: false})
+			case "include-main":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-main")
+			case "exclude-main":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			case "include-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-vmarg")
+			case "exclude-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			}
+		}
+	}
+	return attacher
+}
+
+func (j *JavaAttacher) addCmdLineDiscoveryRule(regexS string, isIncludeRule bool, argumentName string) {
+	regex, err := regexp.Compile(regexS)
+	if err != nil {
+		j.logger.Errorf("invalid regex for the `%v` argument: %v", argumentName, err)
+	}
+	if regex != nil {
+		j.addDiscoveryRule(CmdLineDiscoveryRule{regex: regex, isIncludeRule: isIncludeRule})
+	}
+}
+
+func (j *JavaAttacher) findFirstMatch(jvm *JvmDetails) DiscoveryRule {
+	for _, rule := range j.discoveryRules {
+		if rule.match(jvm) {
+			return rule
+		}
+	}
+	return nil
+}
+
+func (j *JavaAttacher) Run(ctx context.Context) {
+	if !j.enabled {
+		j.logger.Debugf("Java Agent attacher is disabled")
+		return
+	}
+
+	c := make(chan struct{})
+	defer close(c)
+	go func() {
+		for {
+			jvms, err := j.discoverJvmsForAttachment(ctx)
 			if err != nil {
-				return JavaAttacher{}, fmt.Errorf("no java binary found: %v", err)
+				j.logger.Infof("error during JVMs discovery: %v", err)
+			}
+			for _, jvm := range jvms {
+				go func(jvm *JvmDetails) {
+					err := j.attach(ctx, jvm)
+					if err != nil {
+						j.logger.Errorf("error attaching to JVM %v: %v", jvm, err)
+					}
+				}(jvm)
+			}
+			select {
+			case <-ctx.Done():
+				return
+			case <-c:
+				return
+			default:
+				time.Sleep(1 * time.Second)
 			}
-			cfg.JavaBin = bin
 		}
-	} else {
-		// Ensure we're using the correct separators for the system
-		// running apm-server
-		cfg.JavaBin = filepath.FromSlash(cfg.JavaBin)
+	}()
+}
+
+func (j *JavaAttacher) discoverJvmsForAttachment(ctx context.Context) (map[string]*JvmDetails, error) {
```

Comment:
similar to comment below - when you expect that the return value should be used even when an error is returned, I'd rather log the error in this function, and change the signature to not return an error, and return `nil` or `jvms` accordingly

Reply:
I don't think the conversation has been resolved. This is still returning `error` but it seems to log an error and return `nil` in some cases.

I'll unresolve but feel free to solve it again if I am missing something :bow:

Reply:
I did address this by not returning both error and a non-`nil` return value.
However, this and the function called by it (`discoverAllRunningJavaProcesses`) can have two types of errors - such that make them unable to provide a return value and such that don't. I want this function to return errors (with `nil` return value) of the first type and only log errors of the second type.